js_modules/dagster-ui/packages/ui-core/src/asset-graph/util.ts

```diff
@@ -0,0 +1,80 @@
+import {AssetKeyInput} from '../graphql/types';
+
+import {buildAssetNodeStatusContent, StatusCase} from './AssetNodeStatusContent';
+import {LiveDataForNode, tokenForAssetKey} from './Utils';
+
+export function groupAssetsByStatus<
+  T extends {
+    key: AssetKeyInput;
+    definition?: {opNames: string[]; isSource: boolean; isObservable: boolean} | null;
+  },
+>(assets: T[], liveDataByNode: Record<string, LiveDataForNode>) {
+  type StatusesType = {
+    asset: T;
+    status: ReturnType<typeof buildAssetNodeStatusContent>;
+  };
+  const statuses = {
+    successful: [] as StatusesType[],
+    failed: [] as StatusesType[],
+    inprogress: [] as StatusesType[],
+    missing: [] as StatusesType[],
+    loading: false,
+  };
+  if (!Object.keys(liveDataByNode).length) {
+    statuses.loading = true;
+    return statuses;
+  }
+  Object.keys(liveDataByNode).forEach((key) => {
+    const assetLiveData = liveDataByNode[key];
+    const asset = assets.find((asset) => tokenForAssetKey(asset.key) === key);
+    if (!asset?.definition) {
+      console.warn('Expected a definition for asset with key', key);
+      return;
+    }
+    const status = buildAssetNodeStatusContent({
+      assetKey: asset.key,
+      definition: asset.definition,
+      liveData: assetLiveData,
+      expanded: true,
+    });
+    switch (status.case) {
+      case StatusCase.LOADING:
+        statuses.loading = true;
+        break;
+      case StatusCase.SOURCE_OBSERVING:
+        statuses.inprogress.push({asset, status});
+        break;
+      case StatusCase.SOURCE_OBSERVED:
+        statuses.successful.push({asset, status});
+        break;
+      case StatusCase.SOURCE_NEVER_OBSERVED:
+        statuses.missing.push({asset, status});
+        break;
+      case StatusCase.SOURCE_NO_STATE:
+        statuses.missing.push({asset, status});
+        break;
+      case StatusCase.MATERIALIZING:
+        statuses.inprogress.push({asset, status});
+        break;
+      case StatusCase.LATE_OR_FAILED:
+        statuses.failed.push({asset, status});
+        break;
+      case StatusCase.NEVER_MATERIALIZED:
+        statuses.missing.push({asset, status});
+        break;
+      case StatusCase.MATERIALIZED:
+        statuses.successful.push({asset, status});
+        break;
+      case StatusCase.PARTITIONS_FAILED:
+        statuses.failed.push({asset, status});
+        break;
+      case StatusCase.PARTITIONS_MISSING:
+        statuses.missing.push({asset, status});
+        break;
+      case StatusCase.PARTITIONS_MATERIALIZED:
+        statuses.successful.push({asset, status});
+        break;
+    }
```

Comment:
This looks good to me, but if we do another refactor of this it'd be really nice to put the "category" of each status into the status case itself. I think our goal here is to group "all the green ones", "all the red ones", etc. but this is a pretty fragile way of doing it because we can't see the colors here.  (I could see a future case where we decide `SOURCE_NEVER_OBSERVED` should be green and the odds we remember to change this code and the count moves into the other bucket seem low :-) )

Reply:
ah yeah that's probably a good idea